signer/utils.go

```diff
@@ -0,0 +1,114 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+)
+
+// GetKeySpec picks up a recommended signing algorithm for given certificate.
+func GetKeySpec(signingCert *x509.Certificate) (KeySpec, error) {
+	var keyspec KeySpec
+	switch key := signingCert.PublicKey.(type) {
+	case *rsa.PublicKey:
+		switch key.Size() {
+		case 256:
+			keyspec = RSA_2048
+		case 384:
+			keyspec = RSA_3072
+		case 512:
+			keyspec = RSA_4096
+		default:
+			return "", UnsupportedSigningKeyError{keyType: "rsa", keyLength: key.Size()}
+		}
+	case *ecdsa.PublicKey:
+		switch key.Curve.Params().BitSize {
+		case 256:
+			keyspec = EC_256
+		case 384:
+			keyspec = EC_384
+		case 521:
+			keyspec = EC_521
+		default:
+			return "", UnsupportedSigningKeyError{keyType: "ecdsa", keyLength: key.Curve.Params().BitSize}
+		}
+	}
+	return keyspec, nil
+}
+
+// GetLocalSignatureProvider returns the LocalSignatureProvider created using given certificates and private key.
+func GetLocalSignatureProvider(certs []*x509.Certificate, pk crypto.PrivateKey) (*LocalSignatureProvider, error) {
+	if len(certs) == 0 {
+		return nil, MalformedArgumentError{param: "certs"}
+	}
+
+	ks, err := GetKeySpec(certs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return &LocalSignatureProvider{
+		key:     pk,
+		certs:   certs,
+		keySpec: ks,
+	}, nil
+}
+
+// LocalSignatureProvider implements SignatureEnvelope's SignatureProvider to facilitate signing using local certificates and private key.
+type LocalSignatureProvider struct {
+	keySpec KeySpec
+	key     crypto.PrivateKey
+	certs  []*x509.Certificate
+}
+
+func (l LocalSignatureProvider) Sign(bytes []byte) ([]byte, []*x509.Certificate, error) {
```

Comment:
nit: Use pointer receiver so that Go doesn't need to copy LocalSignatureProvider's fields

Reply:
Will do in next revision with signature scheme changes.